sympolyc/src/expr.rs

```diff
@@ -0,0 +1,241 @@
+                .all(|(p, _)| primes.contains(p));
+            let monomial_degree = prime_decomposition_of_index
+                .iter()
+                .fold(0, |acc, (_, d)| acc + d);
+            let is_valid_decomposition = is_valid_decomposition && monomial_degree <= D;
+            if is_valid_decomposition {
+                normalized_indices[j] = i;
+                j += 1;
+            }
+        });
+        normalized_indices
+    }
+
+    /// Returns `true` if the polynomial is homoegenous of degree `d`
+    pub fn is_homogeneous(&self) -> bool {
+        let normalized_indices = self.normalized_indices.clone();
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let is_homogeneous = normalized_indices
+            .iter()
+            .zip(self.coeff.clone())
+            .all(|(idx, c)| {
+                let decomposition_of_i = naive_prime_factors(*idx, &mut prime_gen);
+                let monomial_degree = decomposition_of_i.iter().fold(0, |acc, (_, d)| acc + d);
+                !(monomial_degree != D && c != F::zero())
+            });
+        is_homogeneous
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Default for MVPoly<F, N, D> {
+    fn default() -> Self {
+        MVPoly::new()
+    }
+}
+
+// Addition
+impl<F: PrimeField, const N: usize, const D: usize> Add for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+// Multiplication
+impl<F: PrimeField, const N: usize, const D: usize> Mul<MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        let mut acc = HashMap::new();
```

Comment:
is this acc ever usefull ?

Reply:
Yes, it is a cache to avoid recomputing decomposition. Example: 20 (2 * 2 * 5). If 10 has been computed before (which is the case as X1 * X3 is handled before x1^2 * x3, you don't want to recompute the two factors decomposition.

Reply:
I renamed acc in cache in c05812ef828c1c2884ff7b326e55d950d15ff053